File: typescript-workshop/part-4-behavior/tests/helper/InMemoryStockRepository.ts
import { Stock } from '../../src/domain/Stock';
import { StockRepository } from '../../src/repository/StockRepository';

export class InMemoryStockRepository implements StockRepository {
  private readonly stocks = new Map<string, number>();

  findByProductId(productId: string): Stock | null {
    const quantity = this.stocks.get(productId);
    if (quantity === undefined) return null;
    return new Stock(productId, quantity);
  }

  updateQuantity(productId: string, newQuantity: number): void {
    this.stocks.set(productId, newQuantity);
  }

  withAvailableStock(productId: string, quantity: number): this {
    this.stocks.set(productId, quantity);
    return this;
  }

  availableStock(productId: string): number {
    return this.stocks.get(productId) ?? 0;
  }
}

File: csharp-workshop/Part4.Behavior.Tests/Helper/InMemoryStockRepository.cs
using Part4.Behavior.Domain;
using Part4.Behavior.Repository;

namespace Part4.Behavior.Tests.Helper;

public class InMemoryStockRepository : IStockRepository
{
    private readonly Dictionary<string, int> _stock = new();

    public InMemoryStockRepository WithAvailableStock(string productId, int quantity)
    {
        _stock[productId] = quantity;
        return this;
    }

    public int AvailableStock(string productId)
    {
        _stock.TryGetValue(productId, out var qty);
        return qty;
    }

    public Stock? FindByProductId(string productId)
    {
        _stock.TryGetValue(productId, out var qty);
        return new Stock(productId, qty);
    }

    public void UpdateQuantity(string productId, int newQuantity)
    {
        _stock[productId] = newQuantity;
    }
}
